### ISSUE TITLE
Assertion `ticks_now == trace_ticks' failed to hold

### ISSUE THREAD

**Issue by neon12345** [mozilla/rr#1734]
I get this error and don't know rr enough to see what is going wrong.

[FATAL ReplaySession.cc:840:check_ticks_consistency() errno: SUCCESS] 
 (task 7646 (rec:6129) at time 45583)
 -> Assertion `ticks_now == trace_ticks' failed to hold. ticks mismatch for 'SYSCALL: munmap'; expected 4306038, got 4306037

**Comment by rocallahan** [comment]
Does this happen consistently when you re-record and replay?

What's in `rr dump -b -m -p` around the event number quoted in the error log?

**Comment by neon12345** [comment]
Ok I think this is my fault trying to add F_OFD_ file locking commands to fnctl the wrong way. Can you give me a hint what to consider when adding them? I thought to handle them like the other locking commands.

**Comment by rocallahan** [comment]
Yes, that should work. File a PR with your work-in-progress and I'll try to spot the bug? I assume you've written a small testcase that shows this bug?

**Comment by neon12345** [comment]
Sadly it is a huge program with a highly non-deterministic run. I just realized that i have overwritten my changes with the latest master but i will add them again.

**Comment by rocallahan** [comment]
Then the first thing I'd suggest if you suspect your changes is to create a small testcase to test them. I recommend adapting `flock.c`.

**Comment by rocallahan** [comment]
(We'll want that testcase in order to merge `F_OFD_` support, anyway.)

**Comment by neon12345** [comment]
I'm ready but can not run the tests here. Can I use a single PR for all changes?

**Comment by rocallahan** [comment]
Yes, though I'd prefer separate commits for independent changes.

I can run the tests for you. Anything keeping you from running the tests yourself?

**Comment by neon12345** [comment]
I have some failed tests:
95:fcntl_sig
96:fcntl_sig-no-syscallbuf
176:madvise-no-syscallbuf
369:setuid
370:setuid-no-syscallbuf
647:step_thread
648:step_thread-no-syscallbuf
761:fork_exec_info_thr
762:fork_exec_info_thr-no-syscallbuf
763:get_thread_list
764:get_thread_list-no-syscallbuf
943:fcntl_sig-32
944:fcntl_sig-32-no-syscallbuf
1024:madvise-32-no-syscallbuf
1217:setuid-32
1218:setuid-32-no-syscallbuf
1495:step_thread-32
1496:step_thread-32-no-syscallbuf
1609:fork_exec_info_thr-32
1610:fork_exec_info_thr-32-no-syscallbuf
1611:get_thread_list-32
1612:get_thread_list-32-no-syscallbuf

Doesn't look like they are related to the change though.
For the memfd_create change there is also some stuff in Kenos fork.

**Comment by rocallahan** [comment]
Are you running these tests as root?

**Comment by neon12345** [comment]
Indeed but it only changes setuid.

**Comment by neon12345** [comment]
Need to readd the switch case...

**Comment by neon12345** [comment]
I can record/replay my program without errors now what is a huge win. But rr at the moment kind of defeats the purpose of mmaps, Copying all the data around what makes it a bit slow sometimes.

**Comment by rocallahan** [comment]
> Indeed but it only changes setuid.

If you redirect the verbose test output to a file you should then be able to grep out the results for those failing tests. Please paste them somewhere.

> Copying all the data around what makes it a bit slow sometimes.

Yes, if you mmap a lot of data files it will be slow --- unless you put the trace and the data files on a filesystem which supports COW cloning, e.g. btrfs or xfs. In that case rr should optimize those copies using `BTRFS_IOC_CLONE`.

**Comment by nbp** [comment]
I have a similar issue while replaying firefox under rr 4.2.0, rr 4.3.0 fails with a different error message which spew all the different register content.

Here is the error message I got while replaying with `rr -M replay -a`:

```
…
[rr 2172 113952][Child 2172] WARNING: No inner window available!: file /home/nicolas/mozilla/alternate-dev/dom/base/nsGlobalWindow.cpp, line 10026
[rr 2172 113955][Child 2172] WARNING: No inner window available!: file /home/nicolas/mozilla/alternate-dev/dom/base/nsGlobalWindow.cpp, line 10026
[rr 2172 115132]++DOMWINDOW == 2 (0x1822d10) [pid = 2172] [serial = 2] [outer = 0x173b200]
[FATAL /tmp/nix-build-rr-4.2.0.drv-0/rr-4.2.0-src/src/ReplaySession.cc:742:check_ticks_consistency() errno: 0 'Success'] 
 (task 3747 (rec:2091) at time 124280)
 -> Assertion `ticks_now == trace_ticks' failed to hold. ticks mismatch for 'SYSCALL: setsockopt'; expected 8318556, got 8318560
Launch gdb with
  gdb /home/nicolas/.local/share/rr/firefox-25/mmap_1_hardlink_firefox
and attach to the rr debug server with:
  target remote :3747
```

Looking for the dump command as suggested before, and selecting the one with the matching global time, I have:

```
{
  real_time:1440395.697027 global_time:124280, event:`SYSCALL: setsockopt' (state:ENTERING_SYSCALL) tid:2091, ticks:8318556
 rax:0xffffffffffffffda rbx:0x36 rcx:0xffffffffffffffff rdx:0x4 rsi:0x6 rdi:0x30 rbp:0x7f4fd2375250 rsp:0x7f4fd2374fb0 r8:0x4 r9:0xd1 r10:0x7f4fd2375238 r11:0x246 r12:0x0 r13:0x7ffe0c93ce3f r14:0x3 r15:0x7f4fd23750b0 rip:0x70000002 eflags:0x246 cs:0x33 ss:0x2b ds:0x0 es:0x0 fs:0x0 gs:0x0 orig_rax:0x36 fs_base:0x7f4fd2376700 gs_base:0x0
}
```

**Comment by rocallahan** [comment]
@nbp file a new issue? This looks like a generic divergence which could have any cause. And post the last several events from your log, not just the very last one where the divergence was detected. Thanks!!!
